### ISSUE TITLE
Add templates for templates 😅

### ISSUE THREAD

**Issue by yordis** [straw-hat-team/straw_hat_mailer#12]
Following https://github.com/swoosh/swoosh/issues/160

I am not sure if this is a good idea, normally I don't see it in other platforms but I see the use cases, specially that we do not want to update 5 emails templates when some generic footer change. 

Probably create some partials templates (even use another table for this) and use them in the original templates. Using some mustache syntax for render the partial in some spot.

- [ ] Figure out how mustache will work with this

cc: @princemaple @stevedomin @anthonator folks I would love to have your feedbacks on this, right now I am using mustache because I didn't figure out which template engine to use and `eex` you can call whatever you want inside 😅

**Comment by princemaple** [comment]
On top of my head... why not just EEx? It's the most familiar and optimized way of doing templating in Elixir. BTW, have you seen https://github.com/swoosh/swoosh/pull/163 ?

**Comment by yordis** [comment]
@princemaple because I didn't know how to stop the user putting any random Elixir code using EEx. No any other reason outside of my limited knowledge.

**Comment by princemaple** [comment]
Hmm, partials are still written by the developer of the app, aren't they?
Why do you want to prevent that?

**Comment by yordis** [comment]
What do you mean? I am not following. I don't want user to put any Elixir code in the template without whilelabel it first.

let say somebody try to do `:os.get_system_info()` (I don't if exists, just for sake of example)

**Comment by princemaple** [comment]
I probably missed the point. What kind of partials are you talking about? What are the use cases? Are you trying to build an email service that allows users to create custom templates?

**Comment by yordis** [comment]
@princemaple right now the package allow you to have templates on the database so you don't need to have any file in your project and stuff like that.

Based on the swoosh thread, I am thinking on creating some partials so you could render it in the main template, so you could have some `header partial` and `footer partial` and then use it on the `welcome template` for example.

And,

about `eex` I don't know how I could whitelist what the user could put on the template, because `eex` you could put `any elixir expression` on it.

**Comment by princemaple** [comment]
Right. Apparently I completely missed the point. Somehow I assumed this being an alternative implementation of https://github.com/swoosh/swoosh/pull/163

Now. This is a totally different topic. It's more of a general templating issue than an email issue.

Anyway, my 2 cents, assuming mustache syntax is used:

1. Say we are building something that would allow end users to create template partials (email or not). You are going to need a parser that can break down an input string into chunks that are marked by mustache semantics (something like `[{:string, "hello "}, {:var, "name"}, {:if, ..., ...}]`) You will then apply view functions (you will have to write those too) on the marked chunks to resolve the template logic, with some given context (so you can resolve `{:var, "name"}`), i.e. conditionals, loops, etc, to generate io data that is similar to what EEx does. After that, everything should be the same as writing a Phoenix / Plug app. (wow sounds so much easier than actually doing it)

2. If what you are trying to do is for developers, just in a different way of doing it, I'll have to say that it's a bad idea. EEx turns templates into functions that can very efficiently generate io data output. Reading templates from DB and going through the whole parsing and generating output at runtime is a huge impact on the performance.

The swoosh issue is more that removing the Phoenix dependency, so Swoosh can render templates in any Elixir apps. It's kind of irrelevant to this issue.

**Comment by yordis** [comment]
> If what you are trying to do is for developers
Well, for general propose of make easier the email stuff

> Reading templates from DB and going through the whole parsing and generating output at runtime is a huge impact on the performance.

@princemaple this got my attention because most of the modern email providers allow you to have some templating system. Unless you are doing a crazy amount of emails per seconds and still, at that point the idea of this OTP is to overcome whatever situation 😄

**Comment by stevedomin** [comment]
Hi @yordis,

Starting from the beginning, what problem are you trying to solve with that project in the first place?

Trying to understand so that I can better respond.

**Comment by anthonator** [comment]
My template implementation was built specifically to support any template engine. It supports EEx out of the box because EEx is part of the Elixir standard library.

I'm not sure how my implementation would work for your specific use-case but it may be possible. If your template body was mostly static and you wanted to provide custom headers and footers you could do something like this:

```elixir
<%= header(@user) %>

Welcome <%= @user.first_name %>!

I hope you enjoy this app!

<%= footer(@user) %>
```

```elixir
defmodule Yordis.Email do
  use Swoosh.Template, root: "lib/yordis/email/templates"

  import Swoosh.Email

  def send_email(user)
    new()
    |> to(user.email)
    |> from("yordis@iscool.com")
    |> subject("Welcome")
    |> assign(:user, user)
    |> render_to_text("welcome.html")
    |> Yordis.Mailer.deliver()
  end

  def header(user) do
    ... some logic to get the custom header for this user ...
  end

  def footer(user) do
    ... some logic to get the custom footer for this user ...
  end
end
```

If you'd rather use Mustache you could provide your own engine implementation.

Everything I'm doing with templates is a work in progress so this may change. It also may not be what you're looking for but I thought I'd throw it out there.

**Comment by yordis** [comment]
@stevedomin the project right add the capability to have templates saved in the database (like MailChimp, Sendgrid). This allow me to setup toolings like InfusionSoft where you let the user to chose how a Email will look like and you don't need an engineer coding it in the server for you.

Now the issue with this ticket is, allow to have some partials like @anthonator example shows.

@anthonator right now the only issue with `eex` is that I will have to whitelist the whole AST because you can do any Elixir expression inside. I wish I could find the way that I dont have to use Mustache but I don't see this would works

**Comment by anthonator** [comment]
I doubt that any Elixir commands would be run if you used code for partials from a database table. The reason is that whatever is being pulled from the database will be interpreted as string literal. The Elixir code in the partial would have to be eval-ed in order for it to be executed.

If you did want to allow specific Elixir functions to be executed you’d have to do something custom. I think that’s pretty far outside the scope of any templating library. I would imagine it would be possible with Elixir’s AST though. Elixir has a lot of good resources on meta programming. Even a book or two I think. ElixirForum and the Slack channel are good resources as well.

I think EEx would work for what you want to do. If you want to evaluate the partial from your database and only allow certain Elixir functions that’s going to be custom. If you don’t want any Elixir to be evaluated you get that out of the box.

Hopefully this is helpful!

**Comment by yordis** [comment]
@anthonator this is the current situation https://elixirforum.com/t/strawhat-mailer-email-management-with-templating-capability/9630/7?u=ws_ubi unless I am not understand something about `eex` which could be a possibility for sure
